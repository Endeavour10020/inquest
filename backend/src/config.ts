import "./env";

// TODO productionize use a more featured configuration management solution
export const config = {
    server: { port: parseInt(process.env.BACKEND_PORT || "4000") },
    auth: {
        secret: process.env.AUTH_SECRET || "auth_secret",
    },
    session: {
        secret:
            process.env.BACKEND_SESSION_SECRET ||
            "beeN2AsaGeib4taiJahkoo2bShah9ah",
        name: "session_store",
        // maxAge is in millisecond we set it to 3 days
        maxAge: 1000 * 60 * 60 * 24 * 3,
    },
    storage: {
        client: {
            endPoint: "127.0.0.1",
            port: 9000,
            useSSL: false,
            accessKey: "minio",
            secretKey: "minio123",
        },
        bucket: "bucket",
        region: "us-east-1",
    },
};
