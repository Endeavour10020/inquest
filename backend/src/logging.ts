import winston = require("winston");

function createLogger() {
    switch (process.env.NODE_ENV) {
        case "test":
            return winston.createLogger({
                silent: true,
            });
        case "debug":
            return winston.createLogger({
                level: "debug",
                format: winston.format.prettyPrint(),
                defaultMeta: { instance: "backend" },
                transports: [
                    new winston.transports.Console({
                        format: winston.format.simple(),
                    }),
                ],
            });
        case "production":
        /* follow through */
        /* eslint-disable */
        default:
            /* eslint-enable */
            return winston.createLogger({
                level: "info",
                format: winston.format.prettyPrint(),
                defaultMeta: { instance: "backend" },
                transports: [
                    new winston.transports.Console({
                        format: winston.format.simple(),
                    }),
                ],
            });
    }
}

export const logger = createLogger();
