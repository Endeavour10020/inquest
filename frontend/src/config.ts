export const config = {
    traceSet: "default",
};
