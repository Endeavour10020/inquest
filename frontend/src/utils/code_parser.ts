function* allMatches(regex: RegExp, code: string) {
    for (
        let matches = regex.exec(code);
        matches !== null;
        matches = regex.exec(code)
    ) {
        yield matches;
    }
}

export class CodeParser {
    private functionRegex = /^def ([a-zA-Z0-9_]+)/gm;
    private classRegex = /^class ([a-zA-Z0-9_]+)/gm;
    public readonly newLineIdxs: ReadonlyArray<number>;

    constructor(public readonly code: string) {
        this.code = code;
        this.newLineIdxs = Array.from(this.newLines());
    }

    private *newLines() {
        let position = 0;
        while (position >= 0) {
            const newPosition = this.code.indexOf("\n", position);
            if (newPosition >= 0) yield newPosition;
            else {
                break;
            }
            position = newPosition + 1;
        }
    }

    findLine(index: number) {
        const offset = 2;
        let lo = 0; // inclusive
        let hi = this.newLineIdxs.length - 1; // exclusive
        if (index < this.newLineIdxs[lo]) return 1; // index is on the first line
        if (index >= this.newLineIdxs[hi]) return hi + offset; // index is on the last line

        while (hi - lo > 1) {
            const mid = Math.floor((hi + lo) / 2);
            if (index >= this.newLineIdxs[mid]) {
                lo = mid;
            } else {
                hi = mid;
            }
        }

        return lo + offset;
    }

    findPosition(index: number) {
        const row = this.findLine(index);
        let col: number | null;
        if (row === 1) {
            col = index;
        } else {
            col = index - this.newLineIdxs[row] - 1;
        }
        if (col < 0) col = null;
        return {
            row,
            col,
        };
    }

    findFunctions() {
        return Array.from(this.findFunctionsIterator());
    }

    private *findFunctionsIterator() {
        this.functionRegex.lastIndex = 0;
        for (const matches of allMatches(this.functionRegex, this.code)) {
            yield {
                name: matches[1],
                line: this.findLine(matches.index),
            };
        }
    }

    findClasses() {
        return Array.from(this.findClassesIterator());
    }

    private *findClassesIterator() {
        this.classRegex.lastIndex = 0;
        for (const classMatch of allMatches(this.classRegex, this.code)) {
            const methodRegex = /^[ \t]+(async )?def ([a-zA-Z0-9_]+)/gm;
            const root = /\n[a-zA-Z0-9_]+/gm;
            methodRegex.lastIndex = classMatch.index;
            root.lastIndex = classMatch.index + 1;
            const endOfClass: number =
                root.exec(this.code)?.index ?? this.code.length - 1;
            const methods: { name: string; line: number }[] = [];

            for (const methodMatch of allMatches(methodRegex, this.code)) {
                if (methodMatch.index >= endOfClass) continue;
                methods.push({
                    name: methodMatch[2].toString(),
                    line: this.findLine(methodMatch.index),
                });
            }

            yield {
                name: classMatch[1],
                line: this.findLine(classMatch.index),
                methods,
            };
        }
    }
}
