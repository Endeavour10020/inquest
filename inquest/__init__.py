from .app import Blueprint  # noqa: F401
from .hotpatch import hotpatch
