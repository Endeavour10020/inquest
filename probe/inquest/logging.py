import contextlib

# flake8: noqa

_CALL_BACKS = set()


def log(value):
    # pylint: disable=all
    print(value)
    for callback in _CALL_BACKS:
        try:
            callback(value)
        except:
            pass


def add_callback(value):
    _CALL_BACKS.add(value)


def remove_callback(value):
    _CALL_BACKS.remove(value)


@contextlib.contextmanager
def with_callback(callback):
    add_callback(callback)
    try:
        yield None
    finally:
        remove_callback(callback)
