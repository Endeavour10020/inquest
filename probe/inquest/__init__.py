# flake8: noqa
from .runner import ProbeRunner, enable
