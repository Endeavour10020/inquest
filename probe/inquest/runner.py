import asyncio
import inspect
import logging
import threading
from typing import List, Optional, Union

from inquest.comms.client_provider import ClientProvider
from inquest.comms.exception_sender import ExceptionSender
from inquest.comms.heartbeat import Heartbeat
from inquest.comms.log_sender import LogSender
from inquest.comms.module_sender import ModuleSender
from inquest.comms.trace_set_subscriber import TraceSetSubscriber
from inquest.probe import Probe

LOGGER = logging.getLogger(__name__)


# TODO split out logic of probe runner to only be connecting together 1 or 2
#      existing services rather than running the logic itself
# TODO relative imports are no longer possible to use
class ProbeRunner(threading.Thread):
    """
    runs the probe on a separate thread and communicates to the endpoint
    @param package: the package of the calling function (used for resolving relative imports)
    @param trace_set_key: which trace_set_key to subscribe to for commands
    @param send_modules: whether or not to send_file information to the endpoint for
                         visualization purposes
    """

    package: str
    probe: Probe
    trace_set_key: str

    def __init__(
        self,
        *,
        root: str,
        package: str,
        trace_set_key: str,
        host: str,
        port: int,
        glob: Optional[Union[str, List[str]]],
        exclude: Optional[List[str]],
    ):
        super().__init__()
        self.package = package
        self.trace_set_key = trace_set_key
        self.send_modules = glob is not None
        self.endpoint = f"{host}:{port}"
        self.root = root
        self.probe = Probe(root, package)
        self.glob = glob
        self.exclude = exclude

    def client_consumers(self):
        sender = ExceptionSender()
        consumers = [
            TraceSetSubscriber(
                probe=self.probe,
                trace_set_key=self.trace_set_key,
                package=self.package,
                exception_sender=sender,
            ),
            LogSender(
                trace_set_key=self.trace_set_key,
                exception_sender=sender,
            ),
            Heartbeat(),
            sender,
        ]

        if self.send_modules:
            consumers.append(
                ModuleSender(
                    url=f'http://{self.endpoint}/upload',
                    root=self.root,
                    package=self.package,
                    glob=self.glob,
                    exclude=self.exclude,
                )
            )
        return consumers

    def run(self):
        LOGGER.info('inquest daemon is running')
        evloop = asyncio.new_event_loop()
        evloop.run_until_complete(self._run_async())
        LOGGER.info('inquest daemon closed')

    async def _run_async(self):
        url = f'ws://{self.endpoint}/graphql'
        consumers = self.client_consumers()
        async with ClientProvider(
                trace_set_key=self.trace_set_key,
                url=url,
                consumers=consumers,
        ) as provider:
            await provider.main()


# TODO pass the TraceSet Key in as an argument
def enable(
    *,
    root: str,
    host: str = "localhost",
    port: int = 4000,
    glob: Optional[Union[str, List[str]]] = None,
    daemon: bool = True,
    package: Optional[str] = None,
    exclude: Optional[List[str]] = None,
) -> None:
    '''
    runs the probe in a separate thread
    '''

    if package is None:
        frame = inspect.stack()[1]
        mod = inspect.getmodule(frame[0])
        package = mod.__name__
    probe = ProbeRunner(
        root=root,
        package=package,
        trace_set_key="default",
        host=host,
        port=port,
        glob=glob,
        exclude=exclude,
    )
    probe.setName('inquest probe')
    probe.setDaemon(daemon)
    probe.start()


if __name__ == "__main__":
    enable(daemon=False)
