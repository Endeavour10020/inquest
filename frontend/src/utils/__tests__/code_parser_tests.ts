import { CodeParser } from "../code_parser";

it("simple code test", function () {
    const code = `import logging
import logging.config
from time import sleep

from inquest.runner import enable
from inquest.util import LOGGING_CONFIG

# enable logging to see inquest's internal log statemnets
logging.config.dictConfig(LOGGING_CONFIG)
LOGGER = logging.getLogger(__name__)


def work(value):
    return value + 2


def main():
    try:
        enable()
        LOGGER.info("starting the main loop")
        value = 0
        while True:
            value = work(value)
            sleep(1)
    except KeyboardInterrupt:
        LOGGER.info("closing the main loop")


if __name__ == "__main__":
    main()
    `;

    const parser = new CodeParser(code);
    expect(parser.newLineIdxs).toMatchSnapshot();
    expect(parser.findFunctions()).toMatchObject([
        { line: 13, name: "work" },
        { line: 17, name: "main" },
    ]);

    expect(parser.findClasses()).toMatchObject([]);
});

it("first line code test", function () {
    const code = `def work(value):
    return value + 2`;

    const parser = new CodeParser(code);
    expect(parser.newLineIdxs).toMatchSnapshot();
    expect(parser.findFunctions()).toMatchObject([{ line: 1, name: "work" }]);
    expect(parser.findClasses()).toMatchObject([]);
});

it("weird name test", function () {
    const code = `def Work(value):
    return value + 2`;

    const parser = new CodeParser(code);
    expect(parser.newLineIdxs).toMatchSnapshot();
    expect(parser.findFunctions()).toMatchObject([{ line: 1, name: "Work" }]);
    expect(parser.findClasses()).toMatchObject([]);
});

it("multiple lines", function () {
    const code = `def work(value):
def last(haha):
def friction(hehe)
`;

    const parser = new CodeParser(code);
    const res = parser.findFunctions();
    expect(res).toMatchObject([
        { line: 1, name: "work" },
        { line: 2, name: "last" },
        { line: 3, name: "friction" },
    ]);
    expect(parser.findClasses()).toMatchObject([]);
});

it("test simple class", function () {
    const code = `
def last(haha):
    pass

class Test:
    def work(value):
        pass

def friction(hehe)
    pass
`;

    const parser = new CodeParser(code);
    const res = parser.findFunctions();
    expect(res).toMatchObject([
        { line: 2, name: "last" },
        { line: 9, name: "friction" },
    ]);
    expect(parser.findClasses()).toMatchObject([
        {
            name: "Test",
            line: 5,
            methods: [
                {
                    name: "Test.work",
                    line: 6,
                },
            ],
        },
    ]);
});

it("test simple class with indented function", function () {
    const code = `
def last(haha):
    pass

class Test:
    def work(value):
        pass

def friction(hehe)
    pass

    def work(value):
        pass
`;

    const parser = new CodeParser(code);
    const res = parser.findFunctions();
    expect(res).toMatchObject([
        { line: 2, name: "last" },
        { line: 9, name: "friction" },
    ]);
    expect(parser.findClasses()).toMatchObject([
        {
            name: "Test",
            line: 5,
            methods: [
                {
                    name: "Test.work",
                    line: 6,
                },
            ],
        },
    ]);
});

it("test simple class with larger class", function () {
    const code = `
def last(haha):
    pass

class Test:
    def work(value):
        pass

    def work2(value):
        pass

    def __init__(self, value):
        pass

def friction(hehe)
    pass

    def work(value):
        pass
`;

    const parser = new CodeParser(code);
    const res = parser.findFunctions();
    expect(res).toMatchObject([
        { line: 2, name: "last" },
        { line: 15, name: "friction" },
    ]);
    expect(parser.findClasses()).toMatchObject([
        {
            name: "Test",
            line: 5,
            methods: [
                {
                    name: "Test.work",
                    line: 6,
                },
                {
                    name: "Test.work2",
                    line: 9,
                },
                {
                    name: "Test.__init__",
                    line: 12,
                },
            ],
        },
    ]);
});
