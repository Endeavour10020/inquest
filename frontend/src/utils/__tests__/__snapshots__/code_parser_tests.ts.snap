// Jest Snapshot v1, https://goo.gl/fbAQLP

exports["first line code test 1"] = `
Array [
  16,
]
`;

exports["simple code test 1"] = `
Array [
  14,
  36,
  59,
  60,
  94,
  134,
  135,
  193,
  235,
  272,
  273,
  274,
  291,
  312,
  313,
  314,
  326,
  335,
  352,
  398,
  416,
  436,
  468,
  489,
  519,
  564,
  565,
  566,
  593,
  604,
]
`;

exports["weird name test 1"] = `
Array [
  16,
]
`;
